File: lluvia/python/test/core/test_debug_utils.py
import pytest
import time
import lluvia as ll


def test_hasReceivedVulkanWarningMessages():

    devices = ll.getAvailableDevices()
    assert(devices is not None)

    for dev in devices:
        session = ll.createSession(
            enableDebug=True, loadNodeLibrary=None, device=dev)
        assert(session is not None)
        assert(dev == session.deviceDescriptor)

        memory = session.createMemory(
            ll.MemoryPropertyFlagBits.DeviceLocal, pageSize=0)
        assert(memory is not None)
        
        # create a source buffer without the ll.BufferUsageFlagBits.TransferSrc flag bit
        sourceBuffer = memory.createBuffer(1024, [ll.BufferUsageFlagBits.StorageBuffer,
                                                  ll.BufferUsageFlagBits.TransferDst])
        assert(sourceBuffer is not None)

        # // create a destination buffer without the ll.BufferUsageFlagBits.TransferDst flag bit
        destinationBuffer = memory.createBuffer(1024, [ll.BufferUsageFlagBits.StorageBuffer,
                                                       ll.BufferUsageFlagBits.TransferSrc])
        assert(sourceBuffer is not None)

        cmdBuffer = session.createCommandBuffer()
        assert(cmdBuffer is not None)

        cmdBuffer.begin()
        cmdBuffer.copyBuffer(sourceBuffer, destinationBuffer)
        cmdBuffer.end()

        session.run(cmdBuffer)

        assert(session.hasReceivedVulkanWarningMessages())


if __name__ == "__main__":

    raise SystemExit(pytest.main([__file__]))

File: lluvia/cpp/core/test/test_debug_utils.cpp
/**
 * \file test_SessionCreation.cpp
 * \brief test Session creation
 * \copyright 2017, Juan David Adarve. See AUTHORS for more details
 * \license Apache 2.0, see LICENSE for more details
 */

#define CATCH_CONFIG_MAIN
#include "catch2/catch.hpp"

#include <iostream>
#include <vector>

#include "lluvia/core.h"

TEST_CASE("hasReceivedVulkanWarningMessages", "SessionCreationTest")
{

    const auto availableDevices = ll::Session::getAvailableDevices();

    for (auto deviceDesc : availableDevices) {

        std::cout << "ID: " << deviceDesc.id
                  << " type: " << ll::deviceTypeToString(std::forward<ll::DeviceType>(deviceDesc.deviceType))
                  << " name: " << deviceDesc.name << std::endl;

        auto desc = ll::SessionDescriptor().enableDebug(true).setDeviceDescriptor(deviceDesc);
        auto session = ll::Session::create(desc);
        REQUIRE(session != nullptr);

        auto memory = session->createMemory(ll::MemoryPropertyFlagBits::DeviceLocal, 0);
        REQUIRE(memory != nullptr);

        // create a source buffer without the ll::BufferUsageFlagBits::TransferSrc flag bit
        auto sourceBuffer = memory->createBuffer(1024, ll::BufferUsageFlagBits::StorageBuffer | ll::BufferUsageFlagBits::TransferDst);
        REQUIRE(sourceBuffer != nullptr);

        // create a destination buffer without the ll::BufferUsageFlagBits::TransferDst flag bit
        auto destinationBuffer = memory->createBuffer(1024, ll::BufferUsageFlagBits::StorageBuffer | ll::BufferUsageFlagBits::TransferSrc);
        REQUIRE(destinationBuffer != nullptr);

        // Executing the command buffer below should raise a warning message as the buffers are not
        // create with the correct set of usage flags.
        // Expected warnings:
        // https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkCmdCopyBuffer-srcBuffer-00118
        // https://www.khronos.org/registry/vulkan/specs/1.1-extensions/html/vkspec.html#VUID-vkCmdCopyBuffer-dstBuffer-00120
        auto cmdBuffer = session->createCommandBuffer();
        REQUIRE(cmdBuffer != nullptr);
        cmdBuffer->begin();
        cmdBuffer->copyBuffer(*sourceBuffer, *destinationBuffer);
        cmdBuffer->end();

        session->run(*cmdBuffer);

        // use CHECK instead of REQUIRE as I want to run the test for all available devices.
        CHECK(session->hasReceivedVulkanWarningMessages());
    }
}
